src/api/app/services/meta_controller_service/project_updater.rb

```diff
@@ -0,0 +1,30 @@
+module MetaControllerService
+  class ProjectUpdater
+    def initialize(project: nil, request_data: {}, validator_klass: ::MetaControllerService::RequestValidator)
+      @project = project
+      @request_data = request_data
+      @validator = validator_klass.new(project: project, request_data: request_data)
+    end
+
+    def call
+      @validator.call
+      unless @validator.valid?
+        @errors = @validator.errors
+        return
+      end
+
+      Project.transaction do
+        @errors = @project.update_from_xml(@request_data)[:error]
+        @project.store if valid?
+      end
+    end
+
+    def errors
+      @errors.is_a?(Array) ? @errors.join("\n") : @errors
```

Comment:
Why not always having an array? Just instantiate the error instance variable with an empty array and use ``<<`` instead.

Reply:
we can hide this implementation detail out of the controller. So if its ok I would keep it where it is.